torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,352 @@
+        for bucket_idx, param_buckets_tuple in enumerate(zip(*param_buckets)):
+            self.bucket_sizes.append(0)
+            # Now, we transpose again, so we iterate over bucket_elems, but getting tuples
+            # of params from each device.
+            for idx, param_tuple in enumerate(zip(*param_buckets_tuple)):
+                if not param_tuple[0].requires_grad:
+                    continue
+                for p in param_tuple:
+                    self.bucket_map[p] = bucket_idx
+                self.bucket_sizes[bucket_idx] += 1
+
+        self.buckets = [[[] for _ in range(len(self.device_ids))] for _ in range(len(self.bucket_sizes))]
+        # coalesced bucket for only device 0
+        self.buckets_coalesced = [[] for _ in range(len(self.bucket_sizes))]
+
+        # We will always reduce the bucket following the reverse order
+        self.bucket_order = list(range(len(self.bucket_sizes)))
+        self.bucket_order.reverse()
+        self.buckets_to_reduce = copy.copy(self.bucket_order)
+
+        self.buckets_ready = set()
+        self.reduction_works = []
+
+        self.devs_ready = [0 for _ in range(len(self.bucket_sizes))]
+
+        # default stream tracking to launch nccl reduce kernels
+        self.default_streams = []
+        for dev_id in self.device_ids:
+            with torch.cuda.device(dev_id):
+                self.default_streams.append(torch.cuda.current_stream())
+
+        self._register_grad_hooks()
+
+    def __getstate__(self):
+        attrs = copy.copy(self.__dict__)
+        return attrs
+
+    def __setstate__(self, state):
+        super(DistributedDataParallelC10d, self).__setstate__(state)
+        self._register_grad_hooks()
+
+    def forward(self, *inputs, **kwargs):
+        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
+        self._sync_params()
+        if len(self.device_ids) == 1:
+            return self.module(*inputs[0], **kwargs[0])
+        outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
+        return self.gather(outputs, self.output_device)
+
+    def scatter(self, inputs, kwargs, device_ids):
+        return scatter_kwargs(inputs, kwargs, device_ids, dim=self.dim)
+
+    def parallel_apply(self, replicas, inputs, kwargs):
+        return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
+
+    def gather(self, outputs, output_device):
+        return gather(outputs, output_device, dim=self.dim)
+
+    def train(self, mode=True):
+        super(DistributedDataParallelC10d, self).train(mode)
+        for module in self._module_copies[1:]:
+            module.train(mode)
+
+    def _dist_broadcast_coalesced(self, tensors, buffer_size):
+        for tensors in _take_tensors(tensors, buffer_size):
+            flat_tensors = _flatten_dense_tensors(tensors)
+            work = c10d.broadcast(flat_tensors, 0, self.process_group)
+            work.wait()
```

Comment:
nit: there's no reason to capture the `work`. You could just `c10d.broadcast(...).wait()`

Reply:
Ok